### PR TITLE
feat(app): add send feature

### PR DESCRIPTION
**Pull request by pedronauck**
## Tasks

- [x] Adapt components from `Transaction` and `DApp/TransactionRequest` to be able to re-use them
- [x] Add missing components no core module 
- [x] Create all transactions-related screens and feature

https://user-images.githubusercontent.com/2029172/209880638-93d75732-0532-4812-803d-d2c4488780d0.mp4

### CODE REVIEW

File: packages/app/src/systems/Core/utils/machine.ts
Comment (luizstacio) [inline packages/app/src/systems/Core/utils/machine.ts:2]: We should avoid using new libs when they are not needed. also, rambda even been good. It Would be better to have a more known lib or just write the helpers functions needed in this case.

File: packages/app/src/systems/Send/machines/sendMachine.ts
Comment (luizstacio) [inline packages/app/src/systems/Send/machines/sendMachine.ts:233]: This looks like a great abstraction, but we should avoid including it in this PR as it will create a new pattern only used here. Looking at the feature, this kind of thing can create many machines without the same patterns. A better approach is to split the PR into phases.

1. Add the new abstraction
2. Refactor all the machines to use the new abstraction
3. Use the new abstraction here, as all the code base is updated.

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/app/src/systems/Send/machines/sendMachine.ts
Comment (pedronauck) [reply packages/app/src/systems/Send/machines/sendMachine.ts:233]: Can I let this PR with this modifications and then open a new one including them in the other machines @luizstacio ?

File: packages/app/src/systems/Core/utils/machine.ts
Comment (pedronauck) [reply packages/app/src/systems/Core/utils/machine.ts]: ✅

File: packages/app/src/systems/Send/machines/sendMachine.ts
Comment (luizstacio) [reply packages/app/src/systems/Send/machines/sendMachine.ts:233]: In my opinion, we should focus on refactoring it first and after add new abstractions. When doing abstractions we create new patterns to implement things that make current code that can be done or in progress already legacy. A good approach for this kind of thing would be to create an issue and do when is a good time. This would solve the two things as all the code base will follow the pattern, and after all, the code would be refactored.

But I want another opinion from the team, not to be the single arbiter here. @LuizAsFight

Comment (LuizAsFight) [reply packages/app/src/systems/Send/machines/sendMachine.ts:233]: I agree. We should not merge this part because:
- it's orthogonal to the main objective of the issue: Create send transaction screen
- we should not just implement a whole new pattern when we already have ~10 machines created in other way


Could you open one issue and elaborate what will be the benefits of refactor to include a new layer to manage actions, with a bunch of helpers and complex object management? then we can continue talking from there

**Comment by LuizAsFight** [comment]
I'm reviewing this PR code already, but will first mention functional problems I found

Send Screen:
- in first frame of video, we see a blink of “ConfirmSend” screen before showing unlock dialog
- can we ask password to unlock wallet later in the flow? we only need signature for actually sending the transaction right @luizstacio ?  
- transaction errors happening, but no feedback to the user.
  - we should not try to simulate the transaction if we know that:
    - invalid address
      - validate To field using  isB256 and isBech32 methods exposed by TS SDK
    - amount bigger than balance
      - validate InputAmount
      - if we click on “Max”, balance is set but considering fee. can we fake a transaction as soon as the user access the screen, enabling us to show the transaction fee earlier and cover all scenarios?
    - anyway after validation, if error occurs when trying to simulate, we might want to show TxErrors component as feedback
- regarding Transaction Details card.
  - should be always visible with value of faked transaction mentioned in prev topic
  - if we won’t do this transaction fee predict, at least it should be hidden when inputs get clear after getting transaction. Right now it’s still visible showing 0.000 ETH fee.


here's a record showing all topics mentioned:

https://user-images.githubusercontent.com/8636507/207742791-c3603ee2-c192-49eb-a556-01e9eecbee54.mov

File: packages/app/src/systems/Send/machines/sendMachine.ts
Comment (LuizAsFight) [inline packages/app/src/systems/Send/machines/sendMachine.ts:221]: all this events are global in the machine?

looks like they are specific of `idle` state? if yes let's move them to right state

File: packages/app/src/systems/Send/machines/sendMachine.ts
Comment (LuizAsFight) [inline packages/app/src/systems/Send/machines/sendMachine.ts:120]: this `idle` is not actually `idle` as it does a lot of things in substates...

should be same style of `confirming` state, that has a first `idle` substate

**Review by LuizAsFight — CHANGES_REQUESTED** [review]

File: packages/app/src/systems/Asset/components/AssetSelect/AssetSelect.stories.tsx
Comment (pedronauck) [reply packages/app/src/systems/Asset/components/AssetSelect/AssetSelect.stories.tsx:25]: ✅

File: packages/app/src/systems/Asset/components/AssetSelect/AssetSelect.test.tsx
Comment (pedronauck) [reply packages/app/src/systems/Asset/components/AssetSelect/AssetSelect.test.tsx:29]: ✅

File: packages/app/src/systems/Transaction/components/TxErrors/TxErrors.tsx
Comment (pedronauck) [reply packages/app/src/systems/Transaction/components/TxErrors/TxErrors.tsx:32]: ✅

File: packages/app/src/systems/Transaction/components/TxDetails/TxDetails.tsx
Comment (pedronauck) [reply packages/app/src/systems/Transaction/components/TxDetails/TxDetails.tsx:22]: ✅

File: packages/app/src/systems/Transaction/components/TxDetails/TxDetails.tsx
Comment (pedronauck) [reply packages/app/src/systems/Transaction/components/TxDetails/TxDetails.tsx:35]: ✅

File: packages/app/src/systems/Home/components/HomeActions/HomeActions.tsx
Comment (pedronauck) [reply packages/app/src/systems/Home/components/HomeActions/HomeActions.tsx:29]: because right after we need to check `hasBalance` to show it wrapped in a `Tooltip` or not, so to don't repeat the component entirely inside JSX, we can just save it inside a variable and do this:

```jsx
{hasBalance ? (
  sendButton
) : (
  <Tooltip content="You don't have balance to send">{sendButton}</Tooltip>
)}
```

File: packages/app/src/systems/Send/routes.tsx
Comment (pedronauck) [reply packages/app/src/systems/Send/routes.tsx:7]: It's defined as a "section" inside the design. It has a bunch of functionality, components, hooks, etc. The definition of a system isn't tight of where it's, but how it's, how complex and the ability to be able to use this system coupled of other parts of the application. All these specs fit very well for the Send.

File: packages/app/src/systems/Send/components/SendConfirm/SendConfirm.tsx
Comment (pedronauck) [reply packages/app/src/systems/Send/components/SendConfirm/SendConfirm.tsx:21]: ✅

File: packages/app/src/systems/Send/components/SendConfirm/SendConfirm.tsx
Comment (pedronauck) [reply packages/app/src/systems/Send/components/SendConfirm/SendConfirm.tsx:18]: ✅

File: packages/app/src/systems/Send/components/SendSelect/SendSelect.tsx
Comment (pedronauck) [reply packages/app/src/systems/Send/components/SendSelect/SendSelect.tsx:52]: ✅

File: packages/app/src/systems/Send/components/SendSelect/SendSelect.tsx
Comment (pedronauck) [reply packages/app/src/systems/Send/components/SendSelect/SendSelect.tsx:57]: ✅

File: packages/app/src/systems/Send/machines/sendMachine.ts
Comment (pedronauck) [reply packages/app/src/systems/Send/machines/sendMachine.ts:221]: Nope, they need to be global, because if the machine are in the `checking` state getting info about transaction, and the user start to change some of the form values, the machine need to go back to the `idle`

File: packages/app/src/systems/Send/machines/sendMachine.ts
Comment (pedronauck) [reply packages/app/src/systems/Send/machines/sendMachine.ts:120]: ✅

**Review by luizstacio — CHANGES_REQUESTED** [review]
I think just move out of this PR the utility tools for the machine.

File: packages/docs/package.json
Comment (luizstacio) [inline packages/docs/package.json:5]: We should remove the changes from "export" and "build/utils" from this PR as they are orthogonal changes.

Comment (luizstacio) [reply packages/docs/package.json:5]: Also the preview deployment is broken on the current PR.

File: .changeset/warm-chefs-begin.md
Comment (luizstacio) [inline .changeset/warm-chefs-begin.md:1]: We can delete this file as not changes were made to the SDK or types packages.

**Review by luizstacio — CHANGES_REQUESTED** [review]
The current preview is broken on this PR.
Also will be good to hold to use the changes of the PR #365

File: package.json
Comment (tomiiide) [inline package.json:54]: Thank you for this.

File: packages/app/src/systems/Send/machines/sendMachine.ts
Comment (LuizAsFight) [reply packages/app/src/systems/Send/machines/sendMachine.ts]: interesting

**Review by luizstacio — CHANGES_REQUESTED** [review]
I'm first evaluating just the flow;
1. When typing the destination address, the console throws errors that should not be displayed.
Screenshot 2022-12-28 at 00 23 37
2. When clicking to confirm, it goes throw a loading state. This shows the current screen. We should go to a transaction view screen if the correct loaders.
Screenshot 2022-12-28 at 00 24 02
3. The confirmation page should re-use the transaction request screen or its components. The current screen needs to include information like gas fees.
Screenshot 2022-12-28 at 00 24 13
4. After approving the transaction, we can open the [review transaction screen](https://fuels-wallet-gdaskxprw-fuel-labs.vercel.app/storybook/?path=/story/transaction-pages-txview--usage), removing the need for the custom status report screens. 

On the current request transaction flow, the screen loading and review for approval are shown like this;
Screenshot 2022-12-28 at 00 30 56
Screenshot 2022-12-28 at 00 29 02

**Comment by pedronauck** [comment]
@luizstacio I did all changes you mentioned in the last comment _(the video above is simulating slow 3g network)_

https://user-images.githubusercontent.com/2029172/209875784-5257c88d-991b-4627-a2d7-09dbaaf0dbe6.mp4

**Comment by luizstacio** [comment]
> @luizstacio I did all changes you mentioned in the last comment _(the video above is simulating slow 3g network)_
> 
>  Kapture.2022-12-28.at.18.44.58.mp4

Perfect just two more things;
- After approval and sending, the header could be titled sending the transaction.
- On the transaction view after sending, the status is not green, but the tx is already confirmed.
Screenshot 2022-12-30 at 01 28 12

**Review by luizstacio — CHANGES_REQUESTED** [review]
Perfect just two more things;
- After approval and sending, the header could be titled sending the transaction.
- On the transaction view after sending, the status is not green, but the tx is already confirmed.
   Screenshot 2022-12-30 at 01 28 12
- Remove the images that are not needed anymore.

**Review by luizstacio — COMMENTED** [review]
We should display the address using bech32 here, maybe not related to the PR;
Screenshot 2023-01-09 at 20 55 14

File: packages/app/src/systems/Asset/components/AssetSelect/AssetSelect.tsx
Comment (luizstacio) [inline packages/app/src/systems/Asset/components/AssetSelect/AssetSelect.tsx:121]: Use `formatAmount` here.

File: packages/app/src/systems/Asset/components/AssetSelect/AssetSelect.tsx
Comment (luizstacio) [inline packages/app/src/systems/Asset/components/AssetSelect/AssetSelect.tsx:39]: I think we don't need this comment here?

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/app/src/systems/Asset/components/AssetSelect/AssetSelect.tsx
Comment (luizstacio) [inline packages/app/src/systems/Asset/components/AssetSelect/AssetSelect.tsx:61]: Even with state changes is better to keep using react than access dom.

File: packages/app/src/systems/Asset/components/AssetSelect/AssetSelect.tsx
Comment (luizstacio) [inline packages/app/src/systems/Asset/components/AssetSelect/AssetSelect.tsx:40]: I think we should keep the menu with a hard width, reducing the new for dynamic control here. Any issues on the component that requires us to use dynamic width?

File: packages/app/src/systems/Asset/components/AssetSelect/AssetSelect.tsx
Comment (luizstacio) [inline packages/app/src/systems/Asset/components/AssetSelect/AssetSelect.tsx:66]: Maybe break the component in smaller components?

File: packages/app/src/systems/DApp/components/index.tsx
Comment (luizstacio) [inline packages/app/src/systems/DApp/components/index.tsx:2]: Maybe we need to fix the imports here, as it looks bad to export on index a component from other place.

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/app/src/systems/Send/hooks/useSend.tsx
Comment (luizstacio) [inline packages/app/src/systems/Send/hooks/useSend.tsx:50]: xstate service already caches the answer. Why do we need to use callback here?

**Review by luizstacio — CHANGES_REQUESTED** [review]

**Comment by LuizAsFight** [comment]
functional when running locally

- we have this "suggested" address, is this intended?
  - if yes it's a dev only thing?
- final of amount input is cut (notice Max button)

image

**Comment by LuizAsFight** [comment]
https://user-images.githubusercontent.com/8636507/211689475-66efa75c-b970-41c8-b74a-81dda0ed3643.mov

can we validate if amount is bigger than balance before sending the transaction? 
and show feedback similar to 
image

**Comment by LuizAsFight** [comment]
storybook Send page is broken in PR`s preview

image

File: packages/app/src/systems/Transaction/services/transaction.tsx
Comment (LuizAsFight) [inline packages/app/src/systems/Transaction/services/transaction.tsx:28]: I think we should not use things from mock file inside a service, would prefer to just create the `provider` here

File: packages/app/src/systems/Transaction/services/transaction.tsx
Comment (LuizAsFight) [inline packages/app/src/systems/Transaction/services/transaction.tsx:178]: can we try to get the same `txCost` info, but using `tx utils` ? would be better because:

1- `getTransactionCost` is broken, because it uses `calculateTransactionFee`, which is using a fixed `GAS_PRICE_FACTOR`.
2- `tx utils` will become the default way to get this kind of info, and this provider methods will be deprecated

File: packages/app/src/systems/Transaction/services/transaction.tsx
Comment (LuizAsFight) [inline packages/app/src/systems/Transaction/services/transaction.tsx:203]: can we try to get the same `fee` info, but using `tx utils` ? would be better because:

1- `calculateFee` is broken, because it's using a fixed `GAS_PRICE_FACTOR`.
2- `tx utils` will become the default way to get this kind of info, and this provider methods will be deprecated

File: packages/docs/examples/Transfer.tsx
Comment (LuizAsFight) [inline packages/docs/examples/Transfer.tsx:18]: 👏🏻

**Review by LuizAsFight — CHANGES_REQUESTED** [review]

**Review by luizstacio — CHANGES_REQUESTED** [review]
Balance is not refreshing after the send is complete

**Review by luizstacio — DISMISSED** [review]

File: packages/app/src/systems/Send/machines/sendMachine.ts
Comment (LuizAsFight) [inline packages/app/src/systems/Send/machines/sendMachine.ts:123]: can we use tx utils to get the fee instead?

File: packages/app/src/systems/DApp/__mocks__/dapp-transaction.ts
Comment (LuizAsFight) [inline packages/app/src/systems/DApp/__mocks__/dapp-transaction.ts:24]: can we use tx utils to get the fee instead?

**Review by LuizAsFight — CHANGES_REQUESTED** [review]

**Review by LuizAsFight — APPROVED** [review]

**Pull request merged**
